### PR TITLE
Fix url encoding handling for code parameter.

### PR DESCRIPTION
**Pull request by incfly**
Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

Fix https://github.com/istio-ecosystem/authservice/issues/205

### CODE REVIEW

File: src/common/http/http.cc [hunk 1]
```diff
@@ -53,6 +53,23 @@ bool IsFormDataSafeCharacter(const char character) {
...
+// ascii chars. RFC 3986[3] about URI specifies that:
+// If a reserved character is found in a URI component and
+//  no delimiting role is known for that character, then it must be
+//  interpreted as representing the data octet corresponding to that
+//  character's encoding in US-ASCII.
+// [1]
+// https://developers.google.com/identity/protocols/oauth2/openid-connect#sendauthrequest
+// [2] https://datatracker.ietf.org/doc/html/rfc6749#appendix-A.11.
+// [3] https://www.rfc-editor.org/rfc/rfc3986#page-12
+bool IsOIDCCodeSafeCharacter(const char character) {
```
Comment (Shikugawa) [inline src/common/http/http.cc:69]: I think that `http.cc` should be independent from filter implementation. It seems this rule is OIDC specific. Thus we should move this to `oidc_filter.cc` or new file such as `oidc_query_decoder.cc` and pass it to `Http::DecodeQueryData`. For example, I prefer to implement that as follows:

```
class UrlQueryDecoder {
  virtual absl::string_view decode(absl::string_view) = 0;
};
```

```
class OidcUrlQueryDecoder : public UrlQueryDecoder {
    // implement...
};
```

But, we should prioritize to fix this problem, Therefore we should postpone this refactor as TODO.

**Comment by vikaschoudhary16** [comment]
Accommodating just one character, `/`,  will not help for very long I think. As we discussed offline, in my case it was the client(browser) which is passing code in the decoded form. my authz code was having `%2f` in encoded form which in decoded form is `/`. Is not it possible that there is some other combination of letters in encoded form which when decode is some other key than the `/`. 
Instead of handling `/` individually, can we determine if the code is encoded or decode by checking presence of reserved keys and if already decoded, skip decoding.

File: src/common/http/http.cc (hunk 1 above)
Comment (incfly) [reply src/common/http/http.cc:71]: Add a TODO.

I think putting out of http lib is fine (this is not filter lib, but http, btw). but I'm not sure if we have to setup class/inheritance, especially if there's only one method, without actual state to maintain.
Pure util function seems work well and clear.

**Comment by incfly** [comment]
> Accommodating just one character, /, will not help for very long I think. 

There's actually only a few char to be considred. `/`, `:` `=` etc. The `=` would not be considered valid. Also in other identity provider, the issue does not occur, say Auth0. I think it's fine to be conservative and keep track of what characters we allow explicitly. i.e. only add chars when we encounter one.

https://stackoverflow.com/questions/14856241/url-encoding-of-the-state-parameter-for-google-oauth2-gets-decoded-during-redire

File: src/common/http/http.cc (hunk 1 above)
Comment (incfly) [reply src/common/http/http.cc:71]: ping, is that sufficient for now? @Shikugawa

**Comment by vikaschoudhary16** [comment]
> > Accommodating just one character, /, will not help for very long I think.
> 
> There's actually only a few char to be considred. `/`, `:` `=` etc. The `=` would not be considered valid. Also in other identity provider, the issue does not occur, say Auth0. I think it's fine to be conservative and keep track of what characters we allow explicitly. i.e. only add chars when we encounter one.
> 
> https://stackoverflow.com/questions/14856241/url-encoding-of-the-state-parameter-for-google-oauth2-gets-decoded-during-redire

sounds good!

File: src/common/http/http.cc (hunk 1 above)
Comment (Shikugawa) [reply src/common/http/http.cc:71]: Yes. SGTM.

**Review by Shikugawa — APPROVED** [review]

**Pull request merged**